### PR TITLE
assignment 01: CreateUser, EditUser & GetUserByEmail API endpoints (using code-first approach)

### PR DESCRIPTION
**Pull request by ericbolikowski**

### CODE REVIEW

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/api/ok.response.dto.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { ResponseDto } from "./response.dto.abstract";
+
+export class OkResponseDto<T> extends ResponseDto {
+  public readonly success = false;
```
Comment (realraif) [inline 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/api/ok.response.dto.ts]: I got a `success: false` on the success response

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.controller.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import { FailResponseDto } from "../../../../../libs/api/fail.response.dto";
+import { ICreateUserController } from "./contracts/create-user.controller.interface";
+import { CreateUserRequestDto } from "./contracts/create-user.request-dto";
+import { CreateUserResponseDto } from "./contracts/create-user.response-dto";
+import { ICreateUserUseCase } from "./contracts/create-user.use-case.interface";
+
+export class CreateUserController implements ICreateUserController {
+  constructor(private readonly useCase: ICreateUserUseCase) {}
+
+  async invoke(message: any) {
```
Comment (realraif) [inline 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.controller.ts:15]: Methods/functions shouldn’t be longer than 20 lines (and ideally much less)

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.use-case.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+export class CreateUserUseCase implements ICreateUserUseCase {
+  constructor(private readonly repository: () => Promise<IUserRepository>) {}
+
+  @Log({
+    startMessage: "Running CreateUser with input:",
+    logInputValues: true,
+    endMessage: "Done, result:",
+    logReturnValue: true,
+  })
```
Comment (realraif) [inline 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.use-case.ts:16]: I really like how you've handled the logs

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/utils/result.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  public isFailure: boolean;
+  public error?: string;
+  private _errorType?: ErrorType;
+  protected _value?: SuccessType;
+
+  protected constructor(
+    isSuccess: boolean,
+    error?: string,
+    errorType?: ErrorType,
+    value?: SuccessType
```
Comment (realraif) [inline 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/utils/result.ts:21]: What is the point of the `SuccessType` type? Why is it not simply `value`?

**Review by realraif — COMMENTED** [review]
Great job!

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/api/ok.response.dto.ts (hunk 1 above)
Comment (ericbolikowski) [reply 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/api/ok.response.dto.ts]: Nice catch! Corrected.

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/utils/result.ts (hunk 4 above)
Comment (ericbolikowski) [reply 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/utils/result.ts:21]: It's a generic type parameter. Any function that returns a `Result` uses it to specify what it returns when successful.

Comment (ericbolikowski) [reply 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/utils/result.ts:21]: For example useful in `create-user.controller.ts` to see what the `CreateUser` use case returns: a result which either contains a `UserEntity` (ok case) or a `UserRepositoryError` (fail case):

CleanShot 2023-10-03 at 19 07 14@2x

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.use-case.ts (hunk 3 above)
Comment (ericbolikowski) [reply 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.use-case.ts:16]: Thank you! :)

File: 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.controller.ts (hunk 2 above)
Comment (ericbolikowski) [reply 0_From_Code_First_To_First_Principles/1_code_first/assignment/src/modules/user/commands/use-cases/create-user/create-user.controller.ts:15]: Good one! I had a lot of fun refactoring this one.  What do you think of the result?

**Pull request merged**